src/aiai_eval/task.py

```diff
@@ -420,8 +422,164 @@ def _evaluate_spacy(
+                    )
+
+            scores.append(test_itr_scores)
+
+        # If track_carbon_emissions is true append metrics, to correctly log emissions
+        # data. We avoid mutating, so any downstream evaluations will not try to use
+        # these.
+        metric_configs = list(self.task_config.metrics)
+        if self.evaluation_config.track_carbon_emissions:
+            metric_configs.append(EMISSIONS)
+            metric_configs.append(POWER)
+
+        # Log scores
+        all_scores = log_scores(
+            task_name=self.task_config.pretty_name,
+            metric_configs=metric_configs,
+            scores=scores,
+            model_id=model_config.model_id,
+            only_return_log=self.evaluation_config.only_return_log,
+        )
+        return all_scores
+
+    def _evaluate_spacy_single_iteration(
+        self,
+        idx: int,
+        model_config: ModelConfig,
+        tests: Sequence[Dataset],
+    ) -> Union[dict, Exception]:
+        """
+            Args:
+            idx (int):
+                The index of the current iteration.
+            model_dict (dict):
+                The model dictionary, with keys "model" and "tokenizer".
+            tests (list):
+                A list of bootstraped test datasets.
+
+        Returns:
+            dict or Exception:
+                The keys in the dict correspond to the metrics and values
+                the corresponding values.
+        """
+        scores = list()
+        return_scores = dict()
+        try:
+            # Set random seeds to enforce reproducibility of the randomly
+            # initialised weights
+            random.seed(703 + idx)
+            np.random.seed(703 + idx)
+
+            # Reinitialise a new model
+            model_dict = self._load_model(model_config=model_config)
+            model = model_dict["model"]
+
+            # Get iteration data
+            test = tests[idx]
+
+            # Define batch size, which depends on whether we are testing or not
+            batch_size = 2 if self.evaluation_config.testing else 32
+
+            # Start carbon emissions tracking
+            if self.evaluation_config.track_carbon_emissions:
+                self.carbon_tracker.start()
+
+            # Get model predictions
+            model_predictions, labels = self._get_spacy_predictions_and_labels(
+                model=model, dataset=test, batch_size=batch_size
+            )
+
+            # Check if the spaCy model has been trained on the task at hand. If
+            # not, then skip this benchmark.
+            sample_preds = model_predictions[0]
+            pos_ner_test = isinstance(sample_preds, list) and "" in sample_preds
+            dep_test = isinstance(sample_preds[0], list) and "" in sample_preds[0]
+            if pos_ner_test or dep_test:
+                raise ModelNotTrainedForTask(
+                    task=self.task_config.name, framework="spacy"
+                )
```

Comment:
I'm not sure what's going on here. If the model is trained for NER but not DEP, then it seems like we refuse to evaluate it for a NER task?

Reply:
Right. Should probably also be moved to `ner.py`.